akka-stream/src/main/scala/akka/stream/io/Timeouts.scala

```diff
@@ -0,0 +1,222 @@
+    Flow.wrap(new CompletionTimeout[T](system, timeout))
+
+  /**
+   * If the time between two processed elements exceed the provided timeout, the stream is failed.
+   */
+  // FIXME: Where should system come from? (Actually we need a Scheduler and ExecutionContext)
+  def idleTimeout[T](timeout: FiniteDuration, system: ActorSystem): Flow[T, T, Unit] =
+    Flow.wrap(new IdleTimeout[T](system, timeout))
+
+  /**
+   * If the time between two processed elements *in any direction* exceed the provided timeout, the stream is failed.
+   *
+   * There is a difference between this stage and having two idleTimeout Flows assembled into a BidiStage.
+   * If the timout is configured to be 1 seconds, then this stage will not fail even though there are elements flowing
+   * every second in one direction, but no elements are flowing in the other direction. I.e. this stage considers
+   * the *joint* frequencies of the elements in both directions.
+   */
+  // FIXME: Where should system come from? (Actually we need a Scheduler and ExecutionContext)
+  def idleTimeoutBidi[A, B](timeout: FiniteDuration, system: ActorSystem): BidiFlow[A, A, B, B, Unit] =
+    BidiFlow.wrap(new IdleTimeoutBidi[A, B](system, timeout))
+
+  private abstract class LinearTimeoutStage[T](val system: ActorSystem) extends GraphStage[FlowShape[T, T]] {
+    val in = Inlet[T]("in")
+    val out = Outlet[T]("out")
+    val shape = FlowShape(in, out)
+
+    protected abstract class LinearTimeoutStageLogic extends GraphStageLogic {
+      protected var timerCancellable: Option[Cancellable] = None
+
+      setHandler(out, new OutHandler {
+        override def onPull(): Unit = pull(in)
+        override def onDownstreamFinish(): Unit = completeStage()
+      })
+
+      protected def onTimerTick(): Unit
+
+      final override def postStop(): Unit = {
+        super.postStop()
+        timerCancellable.foreach(_.cancel())
+      }
+    }
+
+  }
+
+  private class InitialTimeout[T](_system: ActorSystem, timeout: FiniteDuration) extends LinearTimeoutStage[T](_system) {
+
+    override def createLogic: GraphStageLogic = new LinearTimeoutStageLogic {
+      private var initialHasPassed = false
+
+      setHandler(in, new InHandler {
+        override def onPush(): Unit = {
+          initialHasPassed = true
+          push(out, grab(in))
+        }
+        override def onUpstreamFinish(): Unit = completeStage()
+        override def onUpstreamFailure(ex: Throwable): Unit = failStage(ex)
+      })
+
+      final override protected def onTimerTick(): Unit =
+        if (!initialHasPassed)
+          failStage(new TimeoutException(s"The first element has not yet passed through in $timeout."))
+
+      override def preStart(): Unit = {
+        super.preStart()
+        val callback = getAsyncCallback[Unit]((_) ⇒ onTimerTick())
+        import system.dispatcher
+        timerCancellable = Some(system.scheduler.scheduleOnce(timeout)(callback.invoke(())))
```

Comment:
Yes, exactly. We will add the necessary methods to the Materializer interface itself.